src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+    // Rust hashes are hex digits with an `h` prepended.
+    let is_rust_hash = |s: &str| s.starts_with('h') && s[1..].chars().all(|c| c.is_digit(16));
+
+    // If there's no trailing Rust hash just return the symbol as is.
+    if symbol.len() < RUST_HASH_LENGTH || !is_rust_hash(&symbol[symbol.len() - RUST_HASH_LENGTH..])
+    {
+        return Cow::Borrowed(symbol);
+    }
+
+    // Strip off trailing hash.
+    let mut rest = &symbol[..symbol.len() - RUST_HASH_LENGTH];
+
+    if rest.ends_with("::") {
+        rest = &rest[..rest.len() - 2];
+    }
+
+    if rest.starts_with("_$") {
+        rest = &rest[1..];
+    }
+
+    let mut demangled = String::new();
+
+    while !rest.is_empty() {
+        if rest.starts_with('.') {
+            if let Some('.') = rest[1..].chars().next() {
+                demangled.push_str("::");
+                rest = &rest[2..];
+            } else {
+                demangled.push_str(".");
+                rest = &rest[1..];
+            }
+        } else if rest.starts_with('$') {
+            macro_rules! demangle {
+                ($($pat:expr => $demangled:expr,)*) => ({
+                    $(if rest.starts_with($pat) {
+                        demangled.push_str($demangled);
+                        rest = &rest[$pat.len()..];
+                        } else)*
+                    {
+                        demangled.push_str(rest);
+                        break;
+                    }
+
+                })
+            }
+
+            demangle! {
+                "$SP$" => "@",
+                "$BP$" => "*",
+                "$RF$" => "&",
+                "$LT$" => "<",
+                "$GT$" => ">",
+                "$LP$" => "(",
+                "$RP$" => ")",
+                "$C$" => ",",
+                "$u7e$" => "~",
+                "$u20$" => " ",
+                "$u27$" => "'",
+                "$u3d$" => "=",
+                "$u5b$" => "[",
+                "$u5d$" => "]",
+                "$u7b$" => "{",
+                "$u7d$" => "}",
+                "$u3b$" => ";",
+                "$u2b$" => "+",
+                "$u21$" => "!",
+                "$u22$" => "\"",
+            }
+        } else {
+            let idx = match rest.char_indices().find(|&(_, c)| c == '$' || c == '.') {
+                None => rest.len(),
+                Some((i, _)) => i,
+            };
+            demangled.push_str(&rest[..idx]);
+            rest = &rest[idx..];
+        }
+    }
+
+    Cow::Owned(demangled)
+}
+
+#[cfg(test)]
+pub(crate) mod testing {
+    use std::collections::HashMap;
+    use std::fmt;
+    use std::fs::File;
+    use std::io::Write;
+    use std::io::{self, BufRead, Read};
+    use std::path::{Path, PathBuf};
+    use std::time::{Duration, Instant};
+
+    use libflate::gzip::Decoder;
+
+    use super::*;
+    use crate::collapse::Collapse;
+
+    pub(crate) trait DurationExt {
+        fn as_nanos_compat(&self) -> u128;
+    }
+
+    impl DurationExt for Duration {
```

Comment:
Maybe add a `TODO` comment to this to say to eventually replace it with `as_nanos()`?